store/sheet.go

```diff
@@ -0,0 +1,320 @@
+	if err != nil {
+		return nil, err
+	}
+
+	if err := tx.Commit(); err != nil {
+		return nil, FormatError(err)
+	}
+
+	return sheet, nil
+}
+
+// PatchSheet updates an existing sheet by ID.
+func (s *SheetService) PatchSheet(ctx context.Context, patch *api.SheetPatch) (*api.Sheet, error) {
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return nil, FormatError(err)
+	}
+	defer tx.Rollback()
+
+	sheet, err := patchSheet(ctx, tx, patch)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := tx.Commit(); err != nil {
+		return nil, FormatError(err)
+	}
+
+	return sheet, nil
+}
+
+// FindSheetList retrieves a list of sheet based on find.
+func (s *SheetService) FindSheetList(ctx context.Context, find *api.SheetFind) ([]*api.Sheet, error) {
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return nil, FormatError(err)
+	}
+	defer tx.Rollback()
+
+	list, err := findSheetList(ctx, tx, find)
+	if err != nil {
+		return []*api.Sheet{}, err
+	}
+
+	return list, nil
+}
+
+// FindSheet retrieves a single sheet based on find.
+// Returns ECONFLICT if finding more than 1 matching records.
+func (s *SheetService) FindSheet(ctx context.Context, find *api.SheetFind) (*api.Sheet, error) {
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return nil, FormatError(err)
+	}
+	defer tx.Rollback()
+
+	list, err := findSheetList(ctx, tx, find)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(list) == 0 {
+		return nil, nil
+	} else if len(list) > 1 {
+		return nil, &common.Error{Code: common.Conflict, Err: fmt.Errorf("found %d sheet with filter %+v, expect 1. ", len(list), find)}
+	}
+	return list[0], nil
+}
+
+// DeleteSheet deletes an existing sheet by ID.
+// Returns ENOTFOUND if sheet does not exist.
+func (s *SheetService) DeleteSheet(ctx context.Context, delete *api.SheetDelete) error {
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return FormatError(err)
+	}
+	defer tx.Rollback()
+
+	err = deleteSheet(ctx, tx, delete)
+	if err != nil {
+		return FormatError(err)
+	}
+
+	if err := tx.Commit(); err != nil {
+		return FormatError(err)
+	}
+
+	return nil
+}
+
+// createSheet creates a new sheet.
+func createSheet(ctx context.Context, tx *Tx, create *api.SheetCreate) (*api.Sheet, error) {
+	row, err := tx.QueryContext(ctx, `
+		INSERT INTO sheet (
+			creator_id,
+			updater_id,
+			instance_id,
+			database_id,
+			name,
+			statement,
+			visibility
+		)
+		VALUES (?, ?, ?, ?, ?, ?, ?)
+		RETURNING id, creator_id, created_ts, updater_id, updated_ts, instance_id, database_id, name, statement, visibility
+	`,
+		create.CreatorID,
+		create.CreatorID,
+		create.InstanceID,
+		create.DatabaseID,
+		create.Name,
+		create.Statement,
+		create.Visibility,
+	)
+
+	if err != nil {
+		return nil, FormatError(err)
+	}
+	defer row.Close()
+
+	row.Next()
+	var sheet api.Sheet
+	if err := row.Scan(
+		&sheet.ID,
+		&sheet.CreatorID,
+		&sheet.CreatedTs,
+		&sheet.UpdaterID,
+		&sheet.UpdatedTs,
+		&sheet.InstanceID,
+		&sheet.DatabaseID,
```

Comment:
This can be NULL, use sql.NullInt32{}, check the example in https://sourcegraph.com/search?q=context:%40bytebase+sql.NullInt32%7B%7D&patternType=literal

Reply:
Get